File: examples/fp-verification-demo-python/src/p3_edge_case.py
# P3 — Edge Cases: 部分保护不充分
import subprocess
import threading


class CommandExecutor:
    """P3-01: 黑名单过滤但不充分"""
    
    BLACKLIST = ["rm", "shutdown", "del", "format"]
    
    def exec(self, cmd: str) -> str:
        cmd = cmd.strip()
        for bad in self.BLACKLIST:
            if bad in cmd:
                return "blocked"
        # 仍有绕过风险：编码、路径组合等
        return subprocess.check_output(cmd, shell=True).decode()


class ConfigManager:
    """P3-02: TOCTOU — 检查在锁外"""
    
    def __init__(self):
        self._loaded = False
        self._lock = threading.Lock()
    
    def reload(self):
        if not self._loaded:             # 检查
            with self._lock:             # 加锁
                self._load_config()
                self._loaded = True
    
    def _load_config(self):
        pass

File: examples/fp-verification-demo-java/src/p3_edge_case.java
// P3 — Edge Cases: 部分保护不充分，标记为 suspected

public class p3_edge_case {

    // P3-01: 黑名单过滤但不充分 — suspected
    public String execCommand(String input) {
        String cmd = input.trim();
        String[] blacklist = {"rm", "shutdown", "format"};
        for (String bad : blacklist) {
            if (cmd.contains(bad)) return "blocked";
        }
        return RuntimeWrapper.exec(cmd);  // 仍有绕过风险
    }

    // P3-02: 读取时加锁但 TOCTOU 窗口 — suspected
    private boolean configLoaded = false;
    private Object configLock = new Object();

    public void reloadConfig() {
        if (!configLoaded) {                // 检查
            synchronized (configLock) {     // 加锁
                // 加载配置...
                configLoaded = true;
            }
        }
    }
}

class RuntimeWrapper {
    public static String exec(String cmd) {
        try {
            java.util.Scanner s = new java.util.Scanner(Runtime.getRuntime().exec(cmd).getInputStream());
            return s.useDelimiter("\\A").hasNext() ? s.next() : "";
        } catch (Exception e) { return ""; }
    }
}
